### PR TITLE
nixos/cfssl: use systemd StateDirectory to provision the data directory

### PR DESCRIPTION
**Pull request by aanderse**
###### Motivation for this change
Currently the `StateDirectory` directive refers to an absolute path in this module which is not allowed.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/services/security/cfssl.nix [hunk 1]
```diff
@@ -164,41 +172,46 @@ in {
...
+              (opt "mutual-tls-cn" mutualTlsCn)
+              (opt "mutual-tls-client-key" mutualTlsClientKey)
+              (opt "mutual-tls-client-cert" mutualTlsClientCert)
+              (opt "tls-remote-ca" tlsRemoteCa)
+              (opt "db-config" dbConfig)
+              (opt "loglevel" (toString logLevel))
+            ];
+        }
+        (mkIf (cfg.dataDir == "/var/lib/cfssl") {
+          StateDirectory =  "cfssl";
```
Comment (pennae) [inline nixos/modules/services/security/cfssl.nix:211]: nit:
```suggestion
        (mkIf (cfg.dataDir == opt.dataDir.default) {
          StateDirectory = baseNameOf cfg.dataDir;
```
with `opt = options.services.cfssl` in the toplevel let.

**Review by pennae — APPROVED** [review]
looks reasonable. wonder whether it'd make sense to run `systemd-analyze verify` over all unit files we generate in a system closure.

File: nixos/modules/services/security/cfssl.nix (hunk 1 above)
Comment (aanderse) [reply nixos/modules/services/security/cfssl.nix:211]: +1 on `baseNameOf`, but I'm not sure I want to declare `opts` to only use it once.

Comment (pennae) [reply nixos/modules/services/security/cfssl.nix:211]: anything that avoid duplicating the path and uses a common definition instead is good here imo. `cfg.dataDir == options.services.cfssl.dataDir.default` is also fine, at least that attrpath is checked while the string itself isn't :) or even binding the path in the toplevel let and using it for default and for the comparison.

Comment (aanderse) [reply nixos/modules/services/security/cfssl.nix:211]: Sorry if it wasn't clear - this is what I meant. Thanks for review, I'll fix this up today.

**Comment by aanderse** [comment]
> wonder whether it'd make sense to run `systemd-analyze verify` over all unit files we generate in a system closure.

That is a great idea!

**Pull request merged**